Publish tags and branches af NPM packages

kasperg opened the pull request:
#### What does this PR do?

This allows versions (source and forks) of the design system to produce releases with can be consumed by Node projects. At the moment the most notable project here is the React components.

#### Where should the reviewer start?

I suggest going through the commits commit by commit.

#### How can the reviewer see this in action?

Check out [builds in my fork](https://github.com/kasperg/dpl-design-system/packages/1344743).

It should be possible to in a temporary project through the following approach:

```
npm init
npm login --scope=@kasperg --registry=https://npm.pkg.github.com/
npm install @kasperg/dpl-design-system@npm-publish
```

kasperg committed:
Publish package to local GitHub npm registry

This allows developers of Node projects to include the design system
as a dependency and use the output from the package in a similar way
as they would use other packages containing static assets.

Forks and branches are a regular part of the development workflow. To
support publishing across these we need to modify the package.json
file of the project accordingly. The json package is added to support
this.Changes are kept in GitHub Actions and not committed to make it
easier to stay up to date with upstream.

Update documentation accordingly.

kasperg committed:
Fix bundle compatiblity for OSX

The -t argument is not supported.

We also switch from + to \; to terminate file commands since that 
also seems to cause problems.

kasperg committed:
Build releases of all branches - not just main

This allows other projects to track continual development which has
not been merged yet.

kasperg committed:
Use fixed version of GitHub auto releases action

Running with latest can be dangerous when you do not know how the
maintainer will handle major release bumps.

Tools like Dependabot should help us keep up to date anyway.

kasperg committed:
Mark all dependencies as dev dependencies

Regular dependencies will be added to projects which depend on this
package. However the current dependencies are not actually needed when
using the output of the package at runtime. They are only needed when
working with Storybook for development.

Consequently we can mark all the dependencies as dev dependencies.
This way dependers only get the built CSS, JS and other assets.

spaceo commented on .github/workflows/deployment.yml:
```diff
@@ -38,3 +39,53 @@ jobs:
+          registry-url: "https://npm.pkg.github.com"
+          scope: "@${{ github.repository_owner }}"
+
+      - run: npm ci
+
+      - name: Building
+        run: npm run css:build
+
+      - name: Bundling assets
+        run: ./bundle.sh
```
Could be nice if these steps could be defined in one place, instead of both in "deploy_release" and "deploy_npm".
Not crucial, just a thought.

kasperg replied:
I agree. Unfortunately I do not think GitHub has a good way to do this.

One way to approach this would be to build and bundle in one job and then fan out to different kind of releases. You can either use cache or artifacts to share the bundle. Artifacts are so slow that is frequently faster just to install packages etc. again. Caching should - in theory at least - be ephemeral and thus do not guarantee availability between jobs.

We could also use a [composite action](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action) which can be reused across jobs. However I fear that would be more trouble that it is worth at this point in time.

All in all I think the best approach for now it to keep things as is.

spaceo replied:
👍

spaceo commented on package.json:
```diff
@@ -3,20 +3,6 @@
   "version": "0.1.0",
   "repository": "https://github.com/danskernesdigitalebibliotek/dpl-design-system",
   "dependencies": {
-    "@testing-library/jest-dom": "^5.11.4",
```
Goes without saying that you have tested this move of dependencies? :)

kasperg replied:
Yep :smile:

`npm run dev` still work as expected and releases contain the same files as before.

spaceo replied:
👍

Adamik10 reviewed (approved):
Looks useful :)

spaceo reviewed (commented):
Looks good!
We have to try out the workflow with tags/branches and see how it works.
I have a few comments...

kasperg committed:
Do not mark Chromatic action as failed despite errors

GitHub Actions does not have an easy way to make an action 
conditional on the availability of a secret. This may be the simplest
way to get around this. 

Remote unnecessary comments while we are at it.

kasperg committed:
Add prefix to tags created for automatic releases

Otherwise we end up in situations where branch and tag names overlap.
Git does not like this and will complain about duplicate ref specs
when trying to push to the branch.

We cannot drop the tags entirely. GitHub requires a tag to create a
non-draft reelase.

kasperg commented:
@spaceo 

Chromatic failures caused by missing tokens should be addressed by 474d13c. Apparently this results in the check not appearing in the list when it is failing.

spaceo commented:
LGTM 👍
